feat: Add constants.humanName to allow countries to have custom full name format

Siyasanga opened the pull request:
In order to allow countries to define their own full name format, we are adding a constants.humanName into the `client.csv` translations, which will then be used everywhere where we are displaying system users and citizen's name in the systems.

Addresses https://github.com/opencrvs/opencrvs-core/issues/6830

Siyasanga committed:
Add humanName to constants.ts

To allow countries to have custom ordering for full names

https://github.com/opencrvs/opencrvs-core/issues/6830

Siyasanga committed:
Use constants.humanName on the UserList.tsx comp

To ensure that we get the format from the country-config

https://github.com/opencrvs/opencrvs-core/issues/6830

Zangetsu101 commented on packages/client/src/views/SysAdmin/Team/user/UserList.tsx:
```diff
@@ -553,10 +553,12 @@ function UserListComponent(props: IProps) {
-                ((createNamesMap(user.name)[intl.locale] as string) ||
-                  (createNamesMap(user.name)[LANG_EN] as string))) ||
+              (user.name &&
+                intl.formatMessage(constantsMessages.humanName, {
+                  firstName: user.name[0].firstNames,
+                  middleName: user.name[0].middleName,
+                  lastName: user.name[0].familyName
+                })) ||
```
It would be better to have a utility function similar to the `createNamesMap`, something like this:
```
function internationalizedName(name: HumanName[], locale: Locale): { 
  firstName: string
  middleName: string
  lastName: string
}
```
It would try to find the name in the given locale, falling back to 'en' in case it does not find it

Siyasanga replied:
I think this is where I was getting confused. It looks like the translations are passed through from `IntlProvider` and also pulled from `country-config`, now passing the locale implies that we have access to all different translations and dynamically changes languages. Do you know where I can get this?

Zangetsu101 replied:
You can just pass in `intl.locale` to the function. As for how to resolve that, take a look at the current implementation of the `createNamesMap` function, it should give you an idea.

rikukissa replied:
Do we really support names in multiple languages 🤔 If not, then I'd aim to simplify as much as we can. IMO best would be a component like 

```
<HumanName name={user.name} />
```

component that could use the `useIntl` hook. Passing the intl object as a function parameter feels like a little upside down

Siyasanga replied:
Hi @rikukissa on the system user we only support one language, English and on the Birth and Death forms we looks like we are only ask `First names` and `Surname`. Does this confirm that we don't support name in multiple languages for `Birth`, `Marriage` and `Death` as well?

rikukissa replied:
Yes, I think that's a safe assumption

Siyasanga commented on packages/client/src/i18n/messages/constants.ts:
```diff
@@ -978,6 +979,11 @@ const messagesToDefine: IConstantsMessages = {
     id: 'constants.withinTargetDaysTo1Year'
+  },
+  humanName: {
+    defaultMessage: `{lastName} {middleName} {firstName}`,
```
Thanks @Zangetsu101 for this question. I was concerned about that as well, when I was manually testing this scenario yesterday it seemed to work just fine when all the `placeholders` are passed as field in the `values`, if the value is `falsey`(`null`,`false`,`undefined` except the number 0 which gets rendered as is) for a `key` it gets omitted in the rendered message.

The only time it seemed to break is when one of the  placeholders(`firstName`,`middleName,`lastName`) is missing from the values array which will result in the rendering: `{lastName} {middleName} {lastName}`

I just struggled to write tests to demostrate these scenarios because I couldn't figure out how to set different locales

Zangetsu101 replied:
hm so you're saying even if provide an empty string for the middleName part, it renders properly? I would have guessed that it would render something like this: `John  Doe`(Notice the double space in between) for the given value: `{ firstName: "John", middleName: '', lastName: "Doe" }`

rikukissa replied:
I think a country can achieve the proper format if we default empty values as `null` than can be then matched against in the language string.

https://formatjs.github.io/docs/core-concepts/icu-syntax/

### Happy path

<img width="987" alt="image" src="https://github.com/user-attachments/assets/c8b067ec-1397-48b6-a190-8c2d15a46a4a">

### Middle name missing (extra space)
<img width="988" alt="image" src="https://github.com/user-attachments/assets/f1fa8df4-7614-4b02-b232-ea667437dd66">

### Defining empty values as `null`

```
{first}{middle, select, null {} other { {middle}}} {last}
```

<img width="1010" alt="image" src="https://github.com/user-attachments/assets/33e94604-e449-45b8-a1b7-9bc212d6ab9e">
<img width="992" alt="image" src="https://github.com/user-attachments/assets/46d0d6d3-1457-4b8e-9dfa-da0582c37489">

Siyasanga committed:
Create a utility function getLocalisedname()

To make the name more usable we had to extract the name formatting logic into it's own function.

https://github.com/opencrvs/opencrvs-core/issues/6830

Siyasanga committed:
Use getLocalisedName on the UserList component

https://github.com/opencrvs/opencrvs-core/issues/6830

Siyasanga commented:
Thank for the feedback guys @rikukissa @Zangetsu101, I ended up going with the utility function approach because I needed just the localised name in some instances instead of the React.Node component. Please let me know if you have feedback

Siyasanga committed:
Use getLocalisedName on the UserAudit comp

Replace older logic to get the name which was based on an assumption that we support names in multiple languages

https://github.com/opencrvs/opencrvs-core/issues/6830

Zangetsu101 commented:
I think we can go with this approach. Removing multiple spaces using regex is better as we would have to handle quite a few edge cases otherwise

Siyasanga committed:
refactor: use getLocalisedName() @ InProgress.tsx

We've found cleaner way to make the rendered name customizable for each country through client copy from country-config

https://github.com/opencrvs/opencrvs-core/issues/6830

Siyasanga committed:
Refactor the work queues to use getLocalisedName()

We need to update all the places where a citizen's name is being referenced to show it in the format that the country chooses

https://github.com/opencrvs/opencrvs-core/issues/6830

Siyasanga committed:
Record changes in the CHANGELOG for this PR

https://github.com/opencrvs/opencrvs-core/issues/6830

tahmidrahman-dsi commented:
@Siyasanga is this PR still work in progress or ready for review?

naftis commented:
I think this PR might have broken the E2E's? @Siyasanga 
![image](https://github.com/user-attachments/assets/555bc732-9add-4374-ac0b-60867907bfca)
